src/UCTSearch.cpp

```diff
@@ -39,8 +39,23 @@
 
 using namespace Utils;
 
-UCTSearch::UCTSearch(GameState & g)
-    : m_rootstate(g) {
+std::unique_ptr<UCTNode> UCTSearch::m_root = nullptr;
```

Comment:
This is a bit ugly, and would prevent using multiple UCTSearches. If we want to reuse information without globals, the owner of UCTSearch can also do that, and that is effectively GTP.cpp's execute().

I would suggest to try the following instead: make think() return the UCTNode for the best move. execute() can call get_move on that to get the move as before. It can also keep a local std::unique_ptr var to point to the result of the last search, and then pass that into the UCTSearch constructor (may be null).

You can then do a check if g.get_last_move() == last_search_node_my_new_root.get_move() and if so (maybe) check hashes to make sure. Checking hashes won't be needed if execute() properly clears it's saved UCTNode on new games etc.

Reply:
See main thread comments for discussion.